beacon_node/network/src/beacon_processor/worker/rpc_methods.rs

```diff
@@ -228,54 +262,84 @@ impl<T: BeaconChainTypes> Worker<T> {
         // remove all skip slots
         let block_roots = block_roots.into_iter().flatten().collect::<Vec<_>>();
 
-        let mut blocks_sent = 0;
-        for root in block_roots {
-            if let Ok(Some(block)) = self.chain.store.get_block(&root) {
-                // Due to skip slots, blocks could be out of the range, we ensure they are in the
-                // range before sending
-                if block.slot() >= req.start_slot
-                    && block.slot() < req.start_slot + req.count * req.step
-                {
-                    blocks_sent += 1;
-                    self.send_network_message(NetworkMessage::SendResponse {
-                        peer_id,
-                        response: Response::BlocksByRange(Some(Box::new(block))),
-                        id: request_id,
-                    });
+        // Fetching blocks is async because it may have to hit the execution layer for payloads.
+        executor.spawn(
+            async move {
+                let mut blocks_sent = 0;
+
+                for root in block_roots {
+                    match self.chain.get_block(&root).await {
+                        Ok(Some(block)) => {
+                            // Due to skip slots, blocks could be out of the range, we ensure they
+                            // are in the range before sending
+                            if block.slot() >= req.start_slot
+                                && block.slot() < req.start_slot + req.count * req.step
+                            {
+                                blocks_sent += 1;
+                                self.send_network_message(NetworkMessage::SendResponse {
+                                    peer_id,
+                                    response: Response::BlocksByRange(Some(Box::new(block))),
+                                    id: request_id,
+                                });
+                            }
+                        }
+                        Ok(None) => {
+                            error!(
+                                self.log,
+                                "Block in the chain is not in the store";
+                                "request_root" => ?root
+                            );
+                            break;
+                        }
+                        Err(e) => {
+                            error!(
+                                self.log,
+                                "Error fetching block for peer";
+                                "block_root" => ?root,
+                                "error" => ?e
+                            );
+                            break;
```

Comment:
I've added `break`s here so that if we get any errors we abort the blocks-by-range response. I think this makes more sense than trying to continue and serving non-contiguous blocks. I think it's also closer to the spirit of the spec which states:

> Clients MUST respond with blocks that are consistent from a single chain within the context of the request.

I wanted to run this by you @AgeManning before I go stomping all over your code.

Reply:
Yeah this seems like reasonable logic. Our sync algo can handle partial responses too.